libkbfs/block_types.go

```diff
@@ -190,6 +225,27 @@ func (fb FileBlock) DeepCopy(codec kbfscodec.Codec) (*FileBlock, error) {
 	return &fileBlockCopy, nil
 }
 
+// UpdateHash updates the hash of this FileBlock
+func (fb *FileBlock) UpdateHash() kbfshash.RawDefaultHash {
+	fb.cacheMtx.Lock()
+	defer fb.cacheMtx.Unlock()
+	_, hash := kbfshash.DoRawDefaultHash(fb.Contents)
+	fb.hash = &hash
+	return hash
+}
+
+// UpdateHashIfNil updates the hash of this FileBlock only if
+// the hash is currently nil
+func (fb *FileBlock) UpdateHashIfNil() kbfshash.RawDefaultHash {
+	fb.cacheMtx.Lock()
```

Comment:
Not sure if it's worth it, but you could do something like this at the start of the function to make read contentions less likely:

```go
h := func() *kbfshash.RawDefaultHash {
  fb.cacheMtx.RLock()
  defer fb.cacheMtx.RUnlock()
  return fb.hash
}()

if h != nil {
  return *h
}

fb.cacheMtx.Lock()
...
```